### PR TITLE
Implement a starter bot in purely functional Scala

### PR DESCRIPTION
**Pull request by Quiescent**
I'd like to contribute this bot for those who are interested in seeing what a bot written in a purely functional style in Scala might look like.
I'm primarily doing this so that my colleagues at work can see an example of purely functional Scala in the wild and might be encouraged to take part in the competition using this style of programming.

*It's absolutely not my intention to in any way downplay the contributions made towards the current Scala starter bot.*
The other Scala bot is, in my opinion, perfectly valid.
This contribution is just a different starting point.

I used a slightly different Dockerfile for this build because I had issues getting the other to work (apt-get couldn't find sbt).
I'm willing to help investigate why there were problems and unify the Scala images if that's the preferred route.

Thanks in advance for the review.
(And thanks to the team at Entellect for organising the competition and putting in the effort!  I really enjoy taking part.)

### CODE REVIEW

**Comment by markvrensburg** [comment]
This is cool idea. I will have a thorough look at it a bit later tonight.

File: starter-bots/purely-functional-scala/project/plugins.sbt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+addCompilerPlugin(
+  "org.scalamacros" % "paradise" % "2.1.1" cross CrossVersion.full
+)
```
Comment (markvrensburg) [inline starter-bots/purely-functional-scala/project/plugins.sbt]: I don't think macro paradise is necessary for this bot, unless you are running on older versions of scala. Is there a particular reason you've added it?

File: starter-bots/purely-functional-scala/src/main/scala/Bot.scala [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      case Some(op) => tryToMoveRandomly(state.currentWormId, state.myPlayer, state.map, op)
+      case None     => ZIO.succeed(None)
+    }) map ((randomMove: Option[Decision]) =>
+    (Seq(opponent flatMap tryToShoot(state.currentWormId, state.myPlayer),
+      randomMove,
+      Some(DoNothing)) collect { case Some(x) => x }).head)
+  }
+
+  def startBot(roundNumber: Int): ZIO[Console with Random with FileReader, Any, Unit] = for {
+    round  <- getStrLn map (_.toInt)
```
Comment (markvrensburg) [inline starter-bots/purely-functional-scala/src/main/scala/Bot.scala]: This `round` variable is not being used at the moment.

**Comment by markvrensburg** [comment]
I think because this is another scala bot, it should be able to run under the same conditions as the "imperative" bot. So maybe for completeness you should as the same README and dockerfile here.

What do you think @Mistborn94?

File: starter-bots/purely-functional-scala/build.sbt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+libraryDependencies += "org.scalaz" %% "scalaz-zio" % "1.0-RC4"
+libraryDependencies += "org.typelevel" %% "cats-core" % "1.6.0"
+val circeVersion = "0.10.0"
+
```
Comment (markvrensburg) [inline starter-bots/purely-functional-scala/build.sbt]: Could you add the Scala version you are planning to run on?

**Comment by Mistborn94** [comment]
> 
> 
> I think because this is another scala bot, it should be able to run under the same conditions as the "imperative" bot. So maybe for completeness you should the same README and dockerfile here.
> 
> What do you think @Mistborn94?

I agree on the docker container at least - we want the same container for both. If this new container works for both I am happy using it, but I think @RohanChhipa and @Privolin need to make that call. 

The readmes do not have to be identical, but each bot needs to have a readme. If there is not one in this PR, we can copy the one from the existing scala bot.

**Comment by Quiescent** [comment]
Hi there.  Thanks for the review!

Sorry for the delay in replying.  I’ve been very busy over the last few days.  I’ll try to reply over the weekend; otherwise I’ll have time again on Monday.

File: starter-bots/purely-functional-scala/project/plugins.sbt (hunk 1 above)
Comment (Quiescent) [reply starter-bots/purely-functional-scala/project/plugins.sbt]: You're correct.  I misread the documentation for circe.
Will be removed on next push.

File: starter-bots/purely-functional-scala/src/main/scala/Bot.scala (hunk 2 above)
Comment (Quiescent) [reply starter-bots/purely-functional-scala/src/main/scala/Bot.scala]: Thanks.  Removed in next push.

File: starter-bots/purely-functional-scala/build.sbt (hunk 3 above)
Comment (Quiescent) [reply starter-bots/purely-functional-scala/build.sbt]: Thanks.  Done in the next push.

**Comment by Quiescent** [comment]
Regarding the existing bots Dockerfile.  I don't think that `sbt` is in the standard Ubuntu package repositories.  The [docs](https://www.scala-sbt.org/download.html) indicate that you should add an `apt` source and the key for the server before trying to install `sbt` with `apt-get`.

The existing Dockerfile FTB on my system (NIX OS 18.09, Docker 18.09.2 -- ask if you would like more details).  Here's the output of building that Dockerfile:

```
$ docker build . -t "scala-bot"
Sending build context to Docker daemon  68.67MB
Step 1/5 : FROM entelectchallenge/base:2019
 ---> 764a01da06ed
Step 2/5 : RUN apt-get update -y
 ---> Using cache
 ---> f0a4aab5a2b8
Step 3/5 : RUN apt-get install default-jdk -y
 ---> Using cache
 ---> b576efe83bad
Step 4/5 : RUN apt-get install maven -y
 ---> Using cache
 ---> 36bdbf99f41f
Step 5/5 : RUN apt-get install sbt -y
 ---> Running in f2e23b660d09
Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package sbt
The command '/bin/sh -c apt-get install sbt -y' returned a non-zero code: 100
```

My Dockerfile builds both my own and the other Scala bot.
Did the other Dockerfile build fine on your end?

**Comment by Quiescent** [comment]
@Mistborn94 I've addressed the outstanding issues.
Please note my comment about the Dockerfile and please could you check whether my redresses are satisfactory.  Thanks.

File: starter-bots/purely-functional-scala/src/main/scala/Bot.scala (hunk 2 above)
Comment (Quiescent) [reply starter-bots/purely-functional-scala/src/main/scala/Bot.scala]: Minor correction: I should have said ignored.  It's still necessary to read the round to trigger the next move :)

**Comment by Mistborn94** [comment]
> 
> 
> @Mistborn94 I've addressed the outstanding issues.
> Please note my comment about the Dockerfile and please could you check whether my redresses are satisfactory. Thanks.

Thanks @Quiescent! I will merge this one in and the let @Privolin and @RohanChhipa make the final call on a working container :)

**Pull request merged**